Client/Frontend/Home/BookmarksEditPanel.swift

```diff
@@ -0,0 +1,335 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import UIKit
+import Storage
+import Shared
+import XCGLogger
+
+private let log = Logger.browserLogger
+
+enum DetailPanel: Int {
+    case Title = 0
+    case URL
+    case Folder
+    case _numberOfPanels
+}
+
+struct BookmarksEditPanelUX {
+    private static let LabelLeftRightPadding = 20
+    private static let LabelTopBottomPadding = 10
+    private static let TextFieldTopBottomPadding = 5
+}
+
+class BookmarksEditPanel: SiteTableViewController, HomePanel, UITextFieldDelegate {
+    weak var homePanelDelegate: HomePanelDelegate? = nil
+    var source: BookmarksModel?
+    var parentFolders = [BookmarkFolder]()
+    var bookmarkFolder: BookmarkFolder?
+    var bookmark: BookmarkItem?
+    
+    var titleField: UITextField?
+    var urlField: UITextField?
+    var doneButton: UIButton?
+    
+    lazy var deleteButtonCell = ButtonPanelCell(style: UITableViewCellStyle.Default, reuseIdentifier: "ButtonPanelCell")
+    var deleteButton: UIButton!
+    
+    private let BookmarksEditPanelCellIdentifier = "BookmarksEditPanelCellIdentifier"
+    private let BookmarkFolderEditHeaderViewIdentifier = "BookmarkFolderEditHeaderIdentifier"
+    
+    init() {
```

Comment:
If we're editing a bookmark, we should have a `BookmarksModel` instance from the previous view controller, no? We can pass that into the init call eliminating the optional since it doesn't make sense for the `source` property to ever be nil if that's what were editing.